### PR TITLE
conformance 008

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**
The bloom filter messages still get flagged as Malformed by zcashd. *shrugs*

### CODE REVIEW

File: src/protocol/payload/filter.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                    "Minimum FilterLoad bytes required is {} but only got {}",
+                    NON_FILTER_BYTES, bytes_read
+                ),
+            ));
+        }
+        let filter_bytes = bytes_read - NON_FILTER_BYTES;
+
+        let mut cursor = std::io::Cursor::new(&buffer[..]);
+
+        let mut filter = vec![0; filter_bytes];
```
Comment (niklaslong) [inline src/protocol/payload/filter.rs:93]: Max size for this is stated as 36k bytes, would be good to check this as well.

**Review by niklaslong — COMMENTED** [review]
Made a few suggestions but it looks good overall. Do you still get a `Malformed` response if you try an empty bloom filter (i.e. `0`s in the filter `vec`)?

File: src/protocol/payload/filter.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    fn encode(&self, buffer: &mut Vec<u8>) -> std::io::Result<()> {
+        buffer.write_all(&self.data)
+    }
+
+    fn decode(bytes: &mut std::io::Cursor<&[u8]>) -> std::io::Result<Self>
+    where
+        Self: Sized,
+    {
+        let mut data = Vec::new();
+        bytes.read_to_end(&mut data)?;
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply src/protocol/payload/filter.rs:46]: Indeed - I am not sure how else one would do this? Unfortunately our interface using `Cursor` isn't strict enough to enforce this truth.

[the remark this answers is not in the document]
Comment (niklaslong) [reply src/protocol/payload/filter.rs:46]: > Indeed - I am not sure how else one would do this? Unfortunately our interface using Cursor isn't strict enough to enforce this truth.

Oh no worries, just wanted to mention it 👍

File: src/protocol/payload/filter.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    filter: Vec<u8>,
+    hash_fn_count: u32,
+    tweak: u32,
+    flags: u8,
+}
+
+impl Default for FilterLoad {
+    fn default() -> Self {
+        Self {
+            filter: vec![47; 36],
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply src/protocol/payload/filter.rs:21]: I've tried simply deriving default i.e. having `filter: Vec::new()`, but its still malformed.

I'll leave both with the trivial `#[derive(Default)]` for now.

File: src/protocol/payload/filter.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            tweak: 123,
+            flags: 0,
+        }
+    }
+}
+
+impl Default for FilterAdd {
+    fn default() -> Self {
+        // random byte value
+        Self { data: vec![47; 36] }
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply src/protocol/payload/filter.rs:32]: Will change to `#[derive(Default)]` until we understand what's required.

**Review by niklaslong — APPROVED** [review]

**Pull request merged**